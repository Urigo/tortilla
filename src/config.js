const Handlebars = require('handlebars');
const Paths = require('./paths');
const Renderer = require('./renderer');
const Utils = require('./utils');

let config = {};

if (Utils.exists(Paths.config)) {
  config = require(Paths.config);
}

module.exports = {
  registerCustomTransformations() {
    if (config.render && config.render.transformations) {
      Object.keys(config.render.transformations).forEach((t) => {
        Renderer.registerTransformation(t, 'diffStep', config.render.transformations[t].bind(this, Handlebars));
      });
    }
  },
  getBlacklist() {
    if (config.render && config.render.blacklist) {
      return config.render.blacklist;
    }

    return null;
  },
};
