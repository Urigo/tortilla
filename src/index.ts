export * from './config';
export * from './editor';
export * from './essentials';
export * from './git';
export * from './local-storage';
export * from './local-cache';
export * from './manual';
export * from './package';
export * from './paths';
export * from './rebase';
export * from './release';
export * from './step';
export * from './submodule';
export * from './utils';
export * from './translator';
export * from './renderer';
