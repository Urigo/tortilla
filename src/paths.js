const Path = require('path');
const Utils = require('./utils');

/**
  It is important to use absolute paths and not relative paths since some helpers
  are distributed over several processes whose execution path is not always the same,
  therefore this module was created.
 */

const cache = {};
const resolve = Path.resolve.bind(Path);

const ascii = resolveTree(resolve(__dirname, 'ascii'), {
  views: resolve(__dirname, 'ascii/views'),
});

const renderer = resolveTree(resolve(__dirname, 'renderer'), {
  helpers: resolve(__dirname, 'renderer/helpers'),
  templates: resolve(__dirname, 'renderer/templates'),
});

const translator = resolveTree(resolve(__dirname, 'translator'), {
  translation: resolve(__dirname, 'translator/translation'),
  locales: resolve(__dirname, 'translator/locales'),
});

const tortilla = resolveTree(resolve(__dirname, '..'), {
  editor: resolve(__dirname, 'editor.js'),
  essentials: resolve(__dirname, 'essentials.js'),
  git: resolve(__dirname, 'git.js'),
  rebase: resolve(__dirname, 'rebase.js'),
  initializer: resolve(__dirname, 'initializer.js'),
  localCache: resolve(__dirname, 'local-cache.js'),
  localStorage: resolve(__dirname, 'local-storage.js'),
  manual: resolve(__dirname, 'manual.js'),
  paths: resolve(__dirname, 'paths.js'),
  release: resolve(__dirname, 'release.js'),
  step: resolve(__dirname, 'step.js'),
  utils: resolve(__dirname, 'utils.js'),
  hooks: resolve(__dirname, 'hooks'),
  skeleton: 'https://github.com/Urigo/tortilla-skeleton.git',
  ascii,
  renderer,
  translator,
});

// Makes the root path available in the branches object using a 'resolve()' method
// e.g. ('foo', { bar: 'bar' }) -> { resolve() -> 'foo', bar: 'bar' }
function resolveTree(root, branches) {
  branches = branches || {};

  return Object.keys(branches).reduce((tree, name) => {
    tree[name] = branches[name];
    return tree;
  }, {
    resolve: Path.resolve.bind(Path, root),
  });
}

// Resolves a bunch of paths to a given tortilla project path
function resolveProject(cwd) {
  if (!cwd) { throw TypeError('A project path must be provided'); }

  if (!process.env.TORTILLA_CACHE_DISABLED && cache[cwd]) {
    return cache[cwd];
  }

  const gitHeads = resolveTree(resolve(cwd, '.git/HEAD'), {
    cherryPick: resolve(cwd, '.git/CHERRY_PICK_HEAD'),
    orig: resolve(cwd, '.git/ORIG_HEAD'),
    revert: resolve(cwd, '.git/REVERT_HEAD'),
  });

  const gitMessages = {
    commit: resolve(cwd, '.git/COMMIT_EDITMSG'),
    merge: resolve(cwd, '.git/MERGE_MSG'),
    squash: resolve(cwd, '.git/SQUASH_MSG'),
  };

  const gitRefs = resolveTree(resolve(cwd, '.git/refs'), {
    heads: resolve(cwd, '.git/refs/heads'),
    remotes: resolve(cwd, '.git/refs/remotes'),
    tags: resolve(cwd, '.git/refs/tags'),
  });

  const git = resolveTree(resolve(cwd, '.git'), {
    ignore: resolve(cwd, '.gitignore'),
    hooks: resolve(cwd, '.git/hooks'),
    rebaseApply: resolve(cwd, '.git/rebase-apply'),
    rebaseMerge: resolve(cwd, '.git/rebase-merge'),
    heads: gitHeads,
    messages: gitMessages,
    refs: gitRefs,
  });

  const npm = {
    ignore: resolve(cwd, '.npmignore'),
    package: resolve(cwd, 'package.json'),
    modules: resolve(cwd, 'node_modules'),
  };

  const manuals = resolveTree(resolve(cwd, '.tortilla/manuals'), {
    templates: resolve(cwd, '.tortilla/manuals/templates'),
    views: resolve(cwd, '.tortilla/manuals/views'),
  });

  return cache[cwd] = resolveTree(cwd, {
    config: resolve(cwd, '.tortilla/config.js'),
    readme: resolve(cwd, 'README.md'),
    renovate: resolve(cwd, 'renovate.json'),
    travis: resolve(cwd, 'travis.yml'),
    locales: resolve(cwd, '.tortilla/locales'),
    storage: resolve(cwd, '.git/.tortilla'),
    manuals,
    tortilla,
    git,
    npm,
    resolveTree,
    resolveProject,
  });
}


module.exports = resolveProject(Utils.cwd());
