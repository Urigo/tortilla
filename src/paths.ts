import * as Path from 'path';
import { Utils } from './utils';

export interface TortillaPaths {
  resolve: () => string;
  resolveTree: (path: string) => object;
  resolveProject: (path: string) => TortillaPaths;
  config: string;
  checkouts: string;
  locales: string;
  readme: string;
  renovate: string;
  storage: string;
  travis: string;
  manuals: {
    resolve: () => string;
    templates: string;
    views: string;
  };
  cli: {
    resolve: () => string;
    tortilla: string;
    tortillaManual: string;
    tortillaRelease: string;
    tortillaStep: string;
    tortillaStrict: string;
    tortillaSubmodule: string;
  };
  git: {
    resolve: () => string;
    ignore: string;
    hooks: string;
    rebaseApply: string;
    rebaseMerge: string;
    heads: {
      resolve: () => string;
      cherryPick: string;
      orig: string;
      revert: string;
    },
    messages: {
      resolve: () => string;
      commit: string;
      merge: string;
      squash: string;
    },
    refs: {
      resolve: () => string;
      heads: string;
      remotes: string;
      tags: string;
    },
  };
  tortillaDir: string;
  tortilla: {
    resolve: () => string;
    dump: string;
    editor: string;
    essentials: string;
    git: string;
    initializer: string;
    localCache: string;
    localStorage: string;
    manual: string;
    package: string;
    paths: string;
    rebase: string;
    release: string;
    step: string;
    submodule: string;
    utils: string;
    hooks: string;
    skeleton: string;
    ascii: {
      resolve: () => string;
      views: string;
    };
    renderer: {
      resolve: () => string;
      helpers: string;
      templates: string;
    };
    translator: {
      resolve: () => string;
      translation: string;
      locales: string;
    };
  };
  npm: {
    resolve: () => string;
    ignore: string;
    package: string;
    modules: string;
  };
}

/**
 It is important to use absolute paths and not relative paths since some helpers
 are distributed over several processes whose execution path is not always the same,
 therefore this module was created.
 */

const cache = {};
const resolve = Path.resolve.bind(Path);

const ascii = resolveTree(resolve(__dirname, 'ascii'), {
  views: resolve(__dirname, 'ascii/views'),
});

const renderer = resolveTree(resolve(__dirname, 'renderer'), {
  helpers: resolve(__dirname, 'renderer/helpers'),
  templates: resolve(__dirname, 'renderer/templates'),
});

const translator = resolveTree(resolve(__dirname, 'translator'), {
  translation: resolve(__dirname, 'translator/translation'),
  locales: resolve(__dirname, 'translator/locales'),
});

const tortilla = resolveTree(resolve(__dirname, '..'), {
  dump: resolve(__dirname, 'dump.js'),
  editor: resolve(__dirname, 'editor.js'),
  essentials: resolve(__dirname, 'essentials.js'),
  git: resolve(__dirname, 'git.js'),
  initializer: resolve(__dirname, 'initializer.js'),
  localCache: resolve(__dirname, 'local-cache.js'),
  localStorage: resolve(__dirname, 'local-storage.js'),
  manual: resolve(__dirname, 'manual.js'),
  package: resolve(__dirname, 'package.js'),
  paths: resolve(__dirname, 'paths.js'),
  rebase: resolve(__dirname, 'rebase.js'),
  release: resolve(__dirname, 'release.js'),
  step: resolve(__dirname, 'step.js'),
  submodule: resolve(__dirname, 'submodule.js'),
  utils: resolve(__dirname, 'utils.js'),
  hooks: resolve(__dirname, 'hooks'),
  skeleton: 'git@github.com:Urigo/tortilla-skeleton.git',
  ascii,
  renderer,
  translator,
});

const cli = resolveTree(resolve(__dirname, '../cli'), {
  tortilla: resolve(__dirname, '../dist/cli/tortilla.js'),
  tortillaManual: resolve(__dirname, '../dist/cli/tortilla-manual.js'),
  tortillaRelease: resolve(__dirname, '../dist/cli/tortilla-release.js'),
  tortillaStep: resolve(__dirname, '../dist/cli/tortilla-step.js'),
  tortillaStrict: resolve(__dirname, '../dist/cli/tortilla-strict.js'),
  tortillaSubmodule: resolve(__dirname, '../dist/cli/tortilla-submodule.js'),
});

// Makes the root path available in the branches object using a 'resolve()' method
// e.g. ('foo', { bar: 'bar' }) -> { resolve() -> 'foo', bar: 'bar' }
function resolveTree(root, branches): any {
  branches = branches || {};

  return Object.keys(branches).reduce((tree, name) => {
    tree[name] = branches[name];

    return tree;
  }, {
    resolve: Path.resolve.bind(Path, root),
  });
}

// Resolves a bunch of paths to a given tortilla project path
export function resolveProject(cwd: string): TortillaPaths {
  if (!cwd) { throw TypeError('A project path must be provided'); }

  if (!process.env.TORTILLA_CACHE_DISABLED && cache[cwd]) {
    return cache[cwd];
  }

  const gitHeads = resolveTree(resolve(cwd, '.git/HEAD'), {
    cherryPick: resolve(cwd, '.git/CHERRY_PICK_HEAD'),
    orig: resolve(cwd, '.git/ORIG_HEAD'),
    revert: resolve(cwd, '.git/REVERT_HEAD'),
  });

  const gitMessages = {
    commit: resolve(cwd, '.git/COMMIT_EDITMSG'),
    merge: resolve(cwd, '.git/MERGE_MSG'),
    squash: resolve(cwd, '.git/SQUASH_MSG'),
  };

  const gitRefs = resolveTree(resolve(cwd, '.git/refs'), {
    heads: resolve(cwd, '.git/refs/heads'),
    remotes: resolve(cwd, '.git/refs/remotes'),
    tags: resolve(cwd, '.git/refs/tags'),
  });

  const git = resolveTree(resolve(cwd, '.git'), {
    ignore: resolve(cwd, '.gitignore'),
    hooks: resolve(cwd, '.git/hooks'),
    rebaseApply: resolve(cwd, '.git/rebase-apply'),
    rebaseMerge: resolve(cwd, '.git/rebase-merge'),
    heads: gitHeads,
    messages: gitMessages,
    refs: gitRefs,
  });

  const npm = {
    ignore: resolve(cwd, '.npmignore'),
    package: resolve(cwd, 'package.json'),
    modules: resolve(cwd, 'node_modules'),
  };

  const manuals = resolveTree(resolve(cwd, '.tortilla/manuals'), {
    templates: resolve(cwd, '.tortilla/manuals/templates'),
    views: resolve(cwd, '.tortilla/manuals/views'),
  });

  return cache[cwd] = resolveTree(cwd, {
    tortillaDir: resolve(cwd, '.tortilla'),
    config: resolve(cwd, '.tortilla/config.js'),
    checkouts: resolve(cwd, '.tortilla/checkouts.json'),
    locales: resolve(cwd, '.tortilla/locales'),
    readme: resolve(cwd, 'README.md'),
    renovate: resolve(cwd, 'renovate.json'),
    storage: resolve(cwd, '.git/.tortilla'),
    travis: resolve(cwd, '.travis.yml'),
    manuals,
    tortilla,
    cli,
    git,
    npm,
    resolveTree,
    resolveProject,
  });
}

export const Paths: TortillaPaths = resolveProject(Utils.cwd());
