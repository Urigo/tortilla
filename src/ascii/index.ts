import * as Fs from 'fs-extra';
import * as Path from 'path';
import {Paths} from '../paths';

/**
  Responsible for printing ascii art files
 */

// Get string for provided ascii art name
function getAsciiArt(artName) {
  const artFile = `${artName}.txt`;
  const artPath = Path.resolve(Paths.tortilla.ascii.views, artFile);

  return Fs.readFileSync(artPath).toString();
}

// Print for provided ascii art
function printAsciiArt(artName) {
  console.log(getAsciiArt(artName));
}

export const Ascii = {
  get: getAsciiArt,
  print: printAsciiArt,
};
