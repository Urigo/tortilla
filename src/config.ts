import * as Handlebars from 'handlebars';
import { Paths } from './paths';
import { Utils } from './utils';

const Renderer = require('./renderer');

export interface TortillaConfig {
  render?: {
    transformations: any[];
    blacklist: any[];
  }
}

let config: TortillaConfig = {};

if (Utils.exists(Paths.config)) {
  config = require(Paths.config) as TortillaConfig;
}

export function registerCustomTransformations() {
  if (config.render && config.render.transformations) {
    Object.keys(config.render.transformations).forEach((t) => {
      Renderer.registerTransformation(t, 'diffStep', config.render.transformations[t].bind(this, Handlebars));
    });
  }
}

export function getBlacklist() {
  if (config.render && config.render.blacklist) {
    return config.render.blacklist;
  }

  return null;
}

export const Config = config;
