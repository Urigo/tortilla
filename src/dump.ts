import * as Fs from 'fs-extra';
import * as Path from 'path';
import * as ReadlineSync from 'readline-sync';
import * as semver from 'semver';
import * as Tmp from 'tmp';
import { Git } from './git';
import { Paths } from './paths';
import { Release } from './release';
import { Utils } from './utils';

const defaultDumpFileName = 'tutorial.json';
const headEnd = '[//]: # (head-end)\n';
const footStart = '[//]: # (foot-start)\n';

// Dumps tutorial into a JSON file
// Output path defaults to cwd
function dumpProject(out: any = Utils.cwd(), options: any = {}) {
  if (out instanceof Object && !(out instanceof String)) {
    options = out;
    out = Utils.cwd();
  }

  options = {
    filter: options.filter,
    reject: options.reject, ...options};

  // Output path is relative to cwd
  out = Path.resolve(Utils.cwd(), out);

  // If provided output path is a dir assume the dump file should be created inside of it
  if (Utils.exists(out, 'dir')) {
    out = Path.join(out, defaultDumpFileName);
  }

  if (Utils.exists(out, 'file')) {
    options.override = options.override || ReadlineSync.keyInYN([
      'Output path already exists.',
      'Would you like to override it and continue?',
    ].join('\n'));

    if (!options.override) {
      return;
    }
  }

  console.log();
  console.log(`Dumping into ${out}...`);
  console.log();

  // Will recursively ensure dirs as well
  Fs.ensureFileSync(out);

  // Run command once
  const tagNames = Git(['tag', '-l'])
    .split('\n')
    .filter(Boolean);

  let branchNames = tagNames
    .filter((tagName) => {
      return /^[^@]+?@\d+\.\d+\.\d+$/.test(tagName);
    })
    .map((tagName) => {
      return tagName.split('@')[0];
    })
    .reduce((prev, branchName) => {
      if (!prev.includes(branchName)) {
        prev.push(branchName);
      }

      return prev;
    }, []);

  if (options.filter) {
    branchNames = branchNames.filter((branchName) => {
      return options.filter.includes(branchName);
    });
  }

  if (options.reject) {
    branchNames = branchNames.filter((branchName) => {
      return !options.reject.includes(branchName);
    });
  }

  const dump = branchNames.map((branchName) => {
    const historyBranchName = `${branchName}-history`;

    // Run command once
    const releaseVersions = tagNames
      .map((tagName) => {
        return tagName.match(new RegExp(`${branchName}@(\\d+\\.\\d+\\.\\d+)`));
      })
      .filter(Boolean)
      .map(match => match[1])
      .reverse();

    const releases = releaseVersions.map((releaseVersion, releaseIndex) => {
      const prevReleaseVersion = releaseVersions[releaseIndex + 1];
      const tagName = `${branchName}@${releaseVersion}`;
      const tagRevision = Git(['rev-parse', tagName]);

      const historyRevision = Git([
        'log', historyBranchName, `--grep=^${tagName}:`, '--format=%H',
      ]).split('\n')
        .filter(Boolean)
        .pop();

      // Instead of printing diff to stdout we will receive it as a buffer
      const changesDiff = Release.diff(prevReleaseVersion, releaseVersion, null, {
        branch: branchName,
        pipe: true
      });

      const manuals = Fs.readdirSync(Paths.manuals.views).sort(Utils.naturalSort).map((manualName, stepIndex) => {
        const format = '%H %s';
        let stepLog;
        let manualPath;

        // Step
        if (stepIndex) {
          manualPath = Path.resolve(Paths.manuals.views, manualName);
          stepLog = Git([
            'log', branchName, `--grep=^Step ${stepIndex}:`, `--format=${format}`,
          ]);
        } else {
          manualPath = Paths.readme;
          stepLog = Git([
            'log', Git.rootHash(branchName), `--format=${format}`,
          ]);
        }

        manualPath = Path.relative(Utils.cwd(), manualPath);
        stepLog = stepLog.split(' ');
        const stepRevision = stepLog.shift();
        const manualTitle = stepLog.join(' ');

        // Removing header and footer, since the view should be used externally on a
        // different host which will make sure to show these two
        let manualView = Git(['show', `${stepRevision}:${manualPath}`]);

        if (manualView.includes(headEnd)) {
          manualView = manualView
            .split(headEnd)
            .slice(1)
            .join(headEnd);
        }

        if (manualView.includes(footStart)) {
          manualView = manualView
            .split(footStart)
            .slice(0, -1)
            .join(footStart);
        }

        manualView = manualView.trim();

        return {
          manualTitle,
          stepRevision,
          manualView,
        };
      });

      return {
        releaseVersion,
        tagName,
        tagRevision,
        historyRevision,
        changesDiff,
        manuals,
      };
    });

    return {
      branchName,
      historyBranchName,
      releases,
    };
  });

  Fs.writeJsonSync(out, dump, { spaces: 2 });

  console.log();
  console.log('Dump finished.');
  console.log();
}

// TODO: Make client calculate the diff on a service worker
// or serve the created HTML file (using SSR)
function diffReleases(dump: string|object, srcTag: string, dstTag: string) {
  let [srcBranchName, srcReleaseVersion] = srcTag.split('@');
  let [dstBranchName, dstReleaseVersion] = dstTag.split('@');

  if (!srcReleaseVersion && !dstReleaseVersion) {
    srcReleaseVersion = srcBranchName;
    dstReleaseVersion = dstBranchName;
    srcBranchName = null;
    dstBranchName = null;
  }

  if (semver.eq(srcReleaseVersion, dstReleaseVersion)) {
    return '';
  }

  // Test result will be used later on
  const reversed = semver.gt(srcReleaseVersion, dstReleaseVersion)

  if (reversed) {
    let temp;

    temp = srcReleaseVersion;
    srcReleaseVersion = dstReleaseVersion;
    dstReleaseVersion = temp;

    temp = srcBranchName;
    srcBranchName = dstBranchName;
    dstBranchName = temp;
  }

  // If an FS path was provided
  if (typeof dump === 'string' || dump instanceof String) {
    // Resolving path relative to cwd
    dump = Path.resolve(Utils.cwd(), dump as string);
    // Parsing JSON
    dump = Fs.readJSONSync(dump);
  }

  const srcDir = buildRelease(dump, srcReleaseVersion, srcBranchName);
  const dstDir = buildRelease(dump, dstReleaseVersion, dstBranchName);

  Fs.removeSync(`${srcDir.name}/.git`);
  Fs.copySync(`${dstDir.name}/.git`, `${srcDir.name}/.git`);

  const diff = Utils.scopeEnv(() => {
    Git(['add', '.']);

    try {
      Git(['commit', '-m', dstReleaseVersion]);
    // No changes were made between releases
    // Probably due to missing versions of submodules
    } catch (e) {
      return '';
    }

    return reversed
      ? Git(['diff', 'HEAD', 'HEAD^'])
      : Git(['diff', 'HEAD^', 'HEAD']);
  }, {
    TORTILLA_CWD: srcDir.name
  });

  srcDir.removeCallback();
  dstDir.removeCallback();

  return diff;
}

function buildRelease(dump: any, releaseVersion: string, branchName?: string) {
  // If no branch was provided, assuming this is a chunk
  const chunk = branchName ? dump.find(c => c.branchName === branchName) : dump;
  const releaseIndex = chunk.releases.findIndex(r => r.releaseVersion === releaseVersion);
  // Most recent release would come LAST
  const releases = chunk.releases.slice(releaseIndex - chunk.releases.length).reverse();
  const diffs = releases.map(r => r.changesDiff).filter(Boolean);
  const dir = Tmp.dirSync({ unsafeCleanup: true });

  Utils.scopeEnv(() => {
    Git(['init']);

    diffs.forEach((diff) => {
      Git(['apply'], {
        input: diff
      });
    });

    Git(['add', '.']);
    Git(['commit', '-m', releaseVersion]);
  }, {
    TORTILLA_CWD: dir.name
  });

  return dir;
}

export const Dump = {
  create: dumpProject,
  diffReleases
};
