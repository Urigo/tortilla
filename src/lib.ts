import * as child_process from 'child_process'
import * as stream from 'stream'
import { Dump } from './dump'
import { Git } from './git'

/*
 lib.ts module provides a runtime access to Tortilla's API through a cluster
 to distribute the work to another core and not to block the main thread
**/

// Will call the function through the worker
const callTortillaFn = (fnName, ...args) => {
  return new Promise((resolve, reject) => {
    // Note that we use a child_process here and not a cluster since we don't wanna
    // fork the main module
    const worker = child_process.fork(__filename, [], {
      env: { ...process.env, TORTILLA_WORKER: '1' }
    })

    worker.send({
      op: 'exec',
      fn: fnName,
      args,
    })

    // On result
    worker.on('message', (msg: any) => {
      if (typeof msg !== 'object') { return }
      if (msg.op !== 'result') { return }
      if (msg.fn !== fnName) { return }

      resolve(msg.payload)
    })

    // On error
    worker.on('message', (msg: any) => {
      if (typeof msg !== 'object') { return }
      if (msg.op !== 'error') { return }
      if (msg.fn !== fnName) { return }

      reject(msg.payload)
    })
  })
}

// Helper function that will simulate a registered user in the git.config
const withUser = (fn) => (...args) => {
  let result = ''
  let nameExists = false
  let emailExists = false

  // Upcoming method requires user.name and user.email to be set
  try {
    // If invocation was successful it means name exists
    try {
      Git(['config', '--global', 'user.name'])
      nameExists = true
    } catch (e) {
      Git(['config', '--global', 'user.name', 'tortilla'])
    }

    // If invocation was successful it means email exists
    try {
      Git(['config', '--global', 'user.email'])
      emailExists = true
    } catch (e) {
      Git(['config', '--global', 'user.email', 'tortilla@tortilla.academy'])
    }

    result = fn(...args)
  }
  finally {
    if (!nameExists) {
      Git(['config', '--global', '--unset', 'user.name'])
    }

    if (!emailExists) {
      Git(['config', '--global', '--unset', 'user.email'])
    }
  }

  return result
}

export const diffReleases = withUser((dump: object, srcRelease: string, destRelease: string) => {
  return Dump.diffReleases(dump, srcRelease, destRelease)
})

// In worker
if (!process.env.TORTILLA_WORKER) {
  // Listen for signals
  process.on('message', (msg: any) => {
    if (msg.op !== 'exec') { return }

    let handler

    switch (msg.fn) {
      case 'diffReleases': {
        handler = diffReleases; break
      }
      // Escape if function name is not supported
      default: return
    }

    try {
      const result = handler(...msg.args)

      process.send({
        op: 'result',
        fn: msg.fn,
        id: msg.id,
        payload: result,
      })
    }
    catch (e) {
      process.send({
        op: 'error',
        fn: msg.fn,
        id: msg.id,
        payload: e,
      })
    }
  })
}
