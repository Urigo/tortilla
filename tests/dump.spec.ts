import { tortillaBeforeAll, tortillaBeforeEach } from './tests-helper';
import './custom-matchers';
import * as Fs from 'fs-extra';
import * as Tmp from 'tmp';
import { Paths } from '../src/paths';

let context: any = {};

describe('Dump', () => {
  beforeAll(() => {
    tortillaBeforeAll.bind(context)();

    context.dumpFile = Tmp.tmpNameSync();

    context.readDumpFile = (parse) => {
      const dumpContent = Fs.readFileSync(context.dumpFile).toString();

      return parse ? JSON.parse(dumpContent) : dumpContent;
    };
  });

  beforeEach(tortillaBeforeEach.bind(context));

  afterEach(function () {
    Fs.removeSync(context.dumpFile);
  });

  it('should dump all branches which has at least a single release', function () {
    const testBranches = ['foo', 'bar', 'baz'];

    testBranches.forEach((branchName) => {
      context.git(['checkout', '-b', branchName]);
      context.tortilla(['release', 'bump', 'minor', '-m', `${branchName} release`]);
      context.git(['checkout', 'master']);
    });

    context.tortilla(['dump', context.dumpFile]);

    expect(context.readDumpFile()).toContainSameContentAsFile('dumps/branches-dump.json');
  });

  it('should dump all releases - sorted by chronological order', function () {
    context.tortilla(['release', 'bump', 'minor', '-m', 'master release 1']);
    context.tortilla(['release', 'bump', 'minor', '-m', 'master release 2']);
    context.tortilla(['release', 'bump', 'minor', '-m', 'master release 3']);
    context.tortilla(['dump', context.dumpFile]);

    expect(context.readDumpFile()).toContainSameContentAsFile('dumps/releases-dump.json');
  });

  it('should dump all manuals - views should have no headers nor footers', function () {
    const comments = ['foo', 'bar', 'baz'];

    comments.forEach((comment, index) => {
      const step = index + 1;

      context.tortilla(['step', 'tag', '-m', comment]);
      context.tortilla(['step', 'edit', step]);
      Fs.writeFileSync(`${Paths.manuals.templates}/step${step}.tmpl`, comment);
      context.tortilla(['manual', 'render', step]);
      context.git(['add', '.']);
      context.git(['commit', '--amend'], { env: { GIT_EDITOR: true } });
      context.git(['rebase', '--continue']);
    });

    context.tortilla(['release', 'bump', 'minor', '-m', 'master release']);
    context.tortilla(['dump', context.dumpFile]);

    expect(context.readDumpFile()).toContainSameContentAsFile('dumps/manuals-dump.json');
  });


  it('should dump all - mixed scenario', function () {
    // Manuals dump
    const comments = ['foo', 'bar', 'baz'];

    comments.forEach((comment, index) => {
      const step = index + 1;

      context.tortilla(['step', 'tag', '-m', comment]);
      context.tortilla(['step', 'edit', step]);
      Fs.writeFileSync(`${Paths.manuals.templates}/step${step}.tmpl`, comment);
      context.tortilla(['manual', 'render', step]);
      context.git(['add', '.']);
      context.git(['commit', '--amend'], { env: { GIT_EDITOR: true } });
      context.git(['rebase', '--continue']);
    });

    const testBranches = ['foo', 'bar', 'baz'];

    testBranches.forEach((branchName) => {
      context.git(['checkout', '-b', branchName]);
      context.tortilla(['release', 'bump', 'minor', '-m', `${branchName} release 1`]);
      context.tortilla(['release', 'bump', 'minor', '-m', `${branchName} release 2`]);
      context.tortilla(['release', 'bump', 'minor', '-m', `${branchName} release 3`]);
      context.git(['checkout', 'master']);
    });

    context.tortilla(['dump', context.dumpFile]);

    expect(context.readDumpFile()).toContainSameContentAsFile('dumps/mixed-dump.json');
  });

  it('should be able to filter branches', function () {
    const testBranches = ['foo', 'bar', 'baz', 'qux'];

    testBranches.forEach((branchName) => {
      context.git(['checkout', '-b', branchName]);
      context.tortilla(['release', 'bump', 'minor', '-m', `${branchName} release`]);
      context.git(['checkout', 'master']);
    });

    context.tortilla(['dump', context.dumpFile, '--filter', 'foo bar baz']);

    expect(context.readDumpFile()).toContainSameContentAsFile('dumps/branches-dump.json');
  });

  it('should be able to reject branches', function () {
    const testBranches = ['foo', 'bar', 'baz', 'qux'];

    testBranches.forEach((branchName) => {
      context.git(['checkout', '-b', branchName]);
      context.tortilla(['release', 'bump', 'minor', '-m', `${branchName} release`]);
      context.git(['checkout', 'master']);
    });

    context.tortilla(['dump', context.dumpFile, '--reject', 'qux']);

    expect(context.readDumpFile()).toContainSameContentAsFile('dumps/branches-dump.json');
  });

  it('should create dirs recursively if output not dir not exist', function () {
    const out = `${context.dumpFile}/foo/bar/baz.json`;
    context.tortilla(['dump', out]);

    expect(context.exists(out, 'file')).toBeTruthy();
  });

  it('should create a tutorial.json file inside dir if already exists', function () {
    Fs.mkdirSync(context.dumpFile);
    context.tortilla(['dump', context.dumpFile]);

    expect(context.exists(`${context.dumpFile}/tutorial.json`, 'file'));
  });
});
