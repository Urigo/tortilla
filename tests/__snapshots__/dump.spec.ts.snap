// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dump create() should dump keywords for each chapter 1`] = `
"[
  {
    \\"repoUrl\\": \\"\\",
    \\"branchName\\": \\"master\\",
    \\"historyBranchName\\": \\"master-history\\",
    \\"releases\\": [
      {
        \\"releaseVersion\\": \\"0.1.0\\",
        \\"releaseDate\\": \\"2020-02-16 09:20:34 -0400\\",
        \\"tagName\\": \\"master@0.1.0\\",
        \\"tagRevision\\": \\"f5b4b104c997b59815075108bcc02b325a0f2cc5\\",
        \\"historyRevision\\": \\"2b90d8dfd7f544ae64960f031003f5a612e3c40d\\",
        \\"changesDiff\\": \\"diff --git a/.gitignore b/.gitignore\\\\nnew file mode 100644\\\\nindex 0000000..5171c54\\\\n--- /dev/null\\\\n+++ b/.gitignore\\\\n@@ -0,0 +1,2 @@\\\\n+node_modules\\\\n+npm-debug.log\\\\n\\\\\\\\ No newline at end of file\\\\ndiff --git a/.tortilla/manuals/templates/root.tmpl b/.tortilla/manuals/templates/root.tmpl\\\\nnew file mode 100644\\\\nindex 0000000..4c874a6\\\\n--- /dev/null\\\\n+++ b/.tortilla/manuals/templates/root.tmpl\\\\n@@ -0,0 +1 @@\\\\n+A newly created Tortilla project. For more information, see https://github.com/Urigo/tortilla.\\\\ndiff --git a/.tortilla/manuals/templates/step1.tmpl b/.tortilla/manuals/templates/step1.tmpl\\\\nnew file mode 100644\\\\nindex 0000000..e69de29\\\\ndiff --git a/.tortilla/manuals/templates/step2.tmpl b/.tortilla/manuals/templates/step2.tmpl\\\\nnew file mode 100644\\\\nindex 0000000..e69de29\\\\ndiff --git a/.tortilla/manuals/templates/step3.tmpl b/.tortilla/manuals/templates/step3.tmpl\\\\nnew file mode 100644\\\\nindex 0000000..e69de29\\\\ndiff --git a/package.json b/package.json\\\\nnew file mode 100644\\\\nindex 0000000..364594f\\\\n--- /dev/null\\\\n+++ b/package.json\\\\n@@ -0,0 +1,11 @@\\\\n+{\\\\n+  \\\\\\"name\\\\\\": \\\\\\"tortilla-project\\\\\\",\\\\n+  \\\\\\"description\\\\\\": \\\\\\"A newly created Tortilla project\\\\\\",\\\\n+  \\\\\\"private\\\\\\": true,\\\\n+  \\\\\\"keywords\\\\\\": [\\\\n+    \\\\\\"tutorial\\\\\\",\\\\n+    \\\\\\"master\\\\\\",\\\\n+    \\\\\\"chapter 3\\\\\\",\\\\n+    \\\\\\"release 1\\\\\\"\\\\n+  ]\\\\n+}\\\\n\\\\\\\\ No newline at end of file\\\\n\\",
        \\"manuals\\": [
          {
            \\"manualTitle\\": \\"Test tortilla project\\",
            \\"stepRevision\\": \\"e4ae78fef1758aeff0562dc861f0d50c18feb4b9\\",
            \\"manualView\\": \\"A newly created Tortilla project. For more information, see https://github.com/Urigo/tortilla.\\",
            \\"keywords\\": []
          },
          {
            \\"manualTitle\\": \\"Step 1: chapter I\\",
            \\"stepRevision\\": \\"145f3460bc80b731a85b685a9a1293a82ba8b904\\",
            \\"manualView\\": \\"\\",
            \\"keywords\\": [
              \\"tutorial\\",
              \\"master\\",
              \\"chapter 1\\",
              \\"release 1\\"
            ]
          },
          {
            \\"manualTitle\\": \\"Step 2: chapter II\\",
            \\"stepRevision\\": \\"33b5b18d943272b8904b2bd933aee233dca7a8bf\\",
            \\"manualView\\": \\"\\",
            \\"keywords\\": [
              \\"tutorial\\",
              \\"master\\",
              \\"chapter 2\\",
              \\"release 1\\"
            ]
          },
          {
            \\"manualTitle\\": \\"Step 3: chapter III\\",
            \\"stepRevision\\": \\"c5ca6d382a0dfc787d96b1cdcdbe7c14a5763359\\",
            \\"manualView\\": \\"\\",
            \\"keywords\\": [
              \\"tutorial\\",
              \\"master\\",
              \\"chapter 3\\",
              \\"release 1\\"
            ]
          }
        ]
      }
    ]
  }
]
"
`;
